File: lecture2.clj

(ns ^{:doc "example natural language interpreter"
      :author "Kyle Richardson"}
    lecture2.core
  (:use [clojure.set])) 


;; set definitions -- evaluates sets directly rather than creating another variable

(def people-studying #{"john" "mary"})
(def people-sleeping #{"bill" "francis"}) 

(def study
  "either determines if an individual is studying, or returns a set"
  (fn
    ;; if there is an argument, see if it is in set
    ([x] (contains?
          people-studying x))
    ;; otherwise return the set
    ([] people-studying))) 

(def sleep
  "determine if an individual is sleeping or return a set"
  (fn
    ([x] (contains?
          people-sleeping x))
    ([] people-sleeping)))

;; negation



;; in clojure, (fn [x] body) is syntactically the same as (lambda (x) body) (in standard lisp) 
;; e.g (lambda x,y : x > 7)(10,3) (python) ==> ((fn [x y] (> x y)) 10 3)

; (def symbol body) associates or binds a symbol ``symbol'' with an anonymous function 

(def neg
  "returns an anonymous, the negation of ``fun''"
  (fn [fun]
    ;; definition of new anonymous function 
    (fn [& args]
      (not
       (apply fun args)))))

(def l-and
  "implementation of conjunction"
  (fn
    ;; written in curried form (takes one argument at a time) 
    ([x]
    (fn [y]
      (fn [f] (and (f x)
                   (f y)))))
    ;; written as a binary relation (takes two arguments 
    ([x y]
    (fn [f] (and (f x)
                 (f y))))))

(def l-or
  "implementation of conjunction"
  (fn
    ;; written in curried form (takes one argument at a time) 
    ([x]
    (fn [y]
      (fn [f] (or (f x)
                  (f y)))))
    ;; written as a binary relation (takes two arguments 
    ([x y]
    (fn [f] (or (f x)
                (f y))))))

(def l-every
  "implementation of ``every'' (as a generalized quantifier)"
  (fn [x,y]
    (let [overlap (count (clojure.set/intersection x y))] 
      (and (> overlap 0)
           (>= overlap (count x))))))

;; when in clojure shell, type (load-file "lecture2.clj")
;; then: (demo)

(def demo
  "show some examples using our interpreter" 
  (fn []
    ;; print out all examles

    (println
     (str
      ; example 1 
      (str (str "john is sleeping: ")
           (str (sleep "john") "\n"))
      
      ;example 2 
      (str (str "bill studies: ")
           (str (study "bill") "\n"))
      
      ;; example 3
      (str (str "bill does not sleep: ")
           (str ((neg study) "bill") "\n"))
      
      ;; example 4
      (str (str "john and mary study: ")
           ;; in curried form
           (str (((l-and "john") "mary") study) "\n"))
      
      ;; example 5
      (str (str "john or mary study: ")
           ;; in binary form 
           (str ((l-or "john" "mary") study) "\n")

      ;; example 6
      (str (str "john or mary do not sleep: ")
           ;; in binary form 
           (str ((l-or "john" "mary") (neg sleep)) "\n"))

      )))))

;; gets called when you do: clj lecture2.clj

(def main- 
  (demo))

File: lecture2.py
## a mini NL interpreter written in a functional style.
##
### run by typing ``python lecture2.py'' on the command line 

## example sets 

students_studying = set(["john","mary"])
students_sleeping = set(["bill","francis"])

## interpretation rules as anonymous functions  

## set operations

study = lambda x : (x in students_studying)
sleep = lambda x : (x in students_sleeping)

## negation

neg = lambda F : (lambda x : not F(x)) ## returns a negated anonymous function

## disjunction ``or'' -- implemented as a binary relation 

log_or = lambda x,y : (lambda  F : F(x) or F(y))

## conjunction ``and'' -- implemented as a binary relation 

log_and = lambda x,y : (lambda F : F(x) and F(y))


## function application

## takes a function and val and returns the function aplied to this value 
fun_application = lambda fun, val : fun(val)


## this gets called when you run python lecture2.py

if __name__ == "__main__":

    ## execute ``John studies''
    print "John studies: %s" % fun_application(study,"john")

    ## execute ``Bill studies''
    print "Bill studies: %s" % fun_application(study,"bill")

    ## execute ``Mary sleeps''
    print "Mary sleeps: %s" % fun_application(sleep,"mary")

    ## execute ``Bill does not studying''
    print "Bill does not study: %s"  %\
       fun_application(fun_application(neg,study),"bill")

    print "john does not sleep: %s"  %\
       fun_application(fun_application(neg,sleep),"john")

    ## execute ``it's not the case that bill is not sleeping
    print "it is not the case that bill is not sleeping: %s" %\
        fun_application(fun_application(neg,fun_application(neg,sleep)),"bill")

    ## execute ``(either) john or bill is studying
    print "either john or bill is studying: %s" %\
        fun_application(log_or("john","bill"),study)

    print "bill and john are studying: %s" %\
      fun_application(log_and("john","bill"),study)

